lib/internal/http2/core.js

```diff
@@ -2696,8 +2696,12 @@ function connect(authority, options, listener) {
   session[kAuthority] = `${options.servername || host}:${port}`;
   session[kProtocol] = protocol;
 
-  if (typeof listener === 'function')
-    session.once('connect', listener);
+  if (typeof listener === 'function') {
+    if (session.connecting)
+      session.once('connect', listener);
+    else
+      listener.call(this, socket);
```

Comment:
The context passed to the `listener` should be `session` rather than `this` from the current scope. Also, please use `Reflect.apply(listener, session, [socket]);` for user-provided callbacks.

Reply:
fixed in 1fee57b30601f351c5b8eb60d2797917c0604a6b